src/api/auth.rs

```diff
@@ -26,94 +31,157 @@ pub struct NoPromptQuery {
     query: web::Query<NoPromptQuery>,
-    id: Identity,
+    id: Option<Identity>,
     session: Session,
     login_manager: web::Data<LoginManager>,
 ) -> Result<HttpResponse, Error> {
-    id.forget();
+    if let Some(id) = id {
+        id.logout();
+    }
     let NoPromptQuery { next, email } = query.into_inner();
     let (url, csrf_token, nonce) = login_manager.login(email);
-    session.insert("csrf_token", csrf_token)?;
-    session.insert("nonce", nonce)?;
+    let login_cookie = LoginCookie { csrf_token, nonce };
     if let Some(next) = next {
         session.insert("next", next)?;
     }
+    let cookie = login_cookie.try_into()?;
     Ok(HttpResponse::TemporaryRedirect()
+        .cookie(cookie)
         .append_header((http::header::LOCATION, url.as_str()))
         .finish())
 }
 
 async fn login(
     query: web::Query<LoginQuery>,
-    id: Identity,
+    id: Option<Identity>,
     session: Session,
     login_manager: web::Data<LoginManager>,
 ) -> Result<HttpResponse, Error> {
-    id.forget();
+    if let Some(id) = id {
+        id.logout();
+    }
     let (url, csrf_token, nonce) = login_manager.login(None);
-    session.insert("csrf_token", csrf_token)?;
-    session.insert("nonce", nonce)?;
+    let login_cookie = LoginCookie { csrf_token, nonce };
     if let Some(next) = query.into_inner().next {
         session.insert("next", next)?;
     }
+
+    let cookie = login_cookie.try_into()?;
     Ok(HttpResponse::TemporaryRedirect()
+        .cookie(cookie)
         .append_header((http::header::LOCATION, url.as_str()))
         .finish())
 }
 
-async fn logout(id: Identity, session: Session, _req: HttpRequest) -> Result<HttpResponse, Error> {
-    id.forget();
+async fn logout(
+    id: Option<Identity>,
+    session: Session,
+    _req: HttpRequest,
+) -> Result<HttpResponse, Error> {
+    if let Some(id) = id {
+        id.logout();
+    }
     session.clear();
+    let cookie = LoginCookie::removal();
     Ok(HttpResponse::Found()
+        .cookie(cookie)
         .append_header((http::header::LOCATION, "/"))
         .finish())
 }
 
 async fn callback(
-    id: Identity,
+    req: HttpRequest,
     pool: web::Data<Pool>,
     session: Session,
     web::Query(q): web::Query<AuthResponse>,
     login_manager: web::Data<LoginManager>,
 ) -> Result<HttpResponse, Error> {
-    let csrf_token: Option<CsrfToken> = session.get("csrf_token")?;
-    let nonce: Option<Nonce> = session.get("nonce")?;
-    let next: String = session.get("next")?.unwrap_or_else(|| String::from("/"));
-    session.clear();
-    match (csrf_token, nonce) {
-        (Some(state), Some(nonce)) if state.secret() == &q.state => {
+    if let Some(login_cookie) = req.cookie(&SETTINGS.auth.login_cookie_name) {
+        let LoginCookie { csrf_token, nonce } = login_cookie.try_into()?;
+        if csrf_token.secret() == &q.state {
             debug!("callback");
+            let next: String = session.get("next")?.unwrap_or_else(|| String::from("/"));
```

Comment:
This doesn't work any more, since the session is now stored in a SameSite Strict cookie, and so isn't sent on the callback request. We always redirect to `/`.

Does `next` need to go into the login cookie?